core/domain/improvements_domain.py

```diff
@@ -135,18 +168,18 @@ def to_dict(self):
                 issue_description: str. The sentence generated by Oppia to
                     describe why the task was created.
                 status: str. Tracks the state/progress of the task entry.
-                resolver_username: str. Username of the user who resolved the
-                    task when status is resolved. Otherwise None.
-                resolver_profile_picture_data_url: str. Profile picture URL of
-                    the user who resolved the task when status is resolved.
-                    Otherwise None.
-                resolved_on_msecs: float. Time in milliseconds since epoch at
-                    which the task was resolved when status is resolved.
-                    Otherwise None.
+                resolver_username: str|None. Username of the user who resolved
+                    the task when status is resolved. Otherwise None.
+                resolver_profile_picture_data_url: str|None. Profile picture
+                    URL of the user who resolved the task when status is
+                    resolved. Otherwise None.
+                resolved_on_msecs: float|datetime.datetime|None. Time in
```

Comment:
This should always be a float (representing msecs). If it's datetime.datetime somewhere, that is wrong and should be changed.

Reply:
[This](https://github.com/oppia/oppia/blob/1f9fe1c065cdc6309d170c7165189b6d68c5bb39/core/domain/improvements_domain.py#L164) expression has type `Union[datetime, None, float]` because `self.resolved_on` can be either None or datetime type and the value of resolved_on_msecs is dependent on `self.resolved_on`. The only way to avoid msecs to have type `datetime` if `self.resolved_on` is not optional which is not possible. So I think this is the only possible way to right.

Reply:
I don't follow; what would cause resolved_on_msecs to have a datetime value?

Reply:
resolved_on_msecs will never be datetime it can either be None or float the [expression](https://github.com/oppia/oppia/blob/1f9fe1c065cdc6309d170c7165189b6d68c5bb39/core/domain/improvements_domain.py#L164)  here is of type `Union[datetime, None, float]` and if I change the type of resolved_on_msecs from `Union[datetime, None, float]` to `optional[float]` then I will get the following error.
![image](https://user-images.githubusercontent.com/37933427/146590824-3bc2a607-4be4-4c68-aaba-2eabdd1a9b5e.png)

Reply:
resolved_on_msecs will never be datetime it can either be None or float the [expression](https://github.com/oppia/oppia/blob/1f9fe1c065cdc6309d170c7165189b6d68c5bb39/core/domain/improvements_domain.py#L164)  here is of type `Union[datetime, None, float]` and if I change the type of resolved_on_msecs from `Union[datetime.datetime, None, float]` to `optional[float]` then I will get the following error.
![image](https://user-images.githubusercontent.com/37933427/146590824-3bc2a607-4be4-4c68-aaba-2eabdd1a9b5e.png)

Reply:
OK, then try to change the expression so that `datetime` isn't a possibility for its type.

Reply:
Done. Thanks, btw that short circuit article was informative.